python/cucim/src/cucim/skimage/morphology/convex_hull.py

```diff
@@ -0,0 +1,315 @@
+from itertools import product
+
+import cupy as cp
+import numpy as np
+
+try:
+    from scipy.spatial import ConvexHull, QhullError
+
+    scipy_available = True
+except ImportError:
+    scipy_available = False
+
+try:
+    from skimage.util import unique_rows
+
+    unique_rows_available = True
+except ImportError:
+    unique_rows_available = False
+
+
+from cucim.skimage._shared.utils import warn
+from cucim.skimage._vendored import ndimage as ndi
+from cucim.skimage.measure._label import label
+from cucim.skimage.measure._regionprops_gpu_utils import _unravel_loop_index
+
+__all__ = [
+    "convex_hull_image",
+    "convex_hull_object",
+]
+
+
+@cp.memoize(for_each_device=True)
+def _offsets_diamond(ndim):
+    offsets = np.zeros((2 * ndim, ndim))
+    for vertex, (axis, offset) in enumerate(product(range(ndim), (-0.5, 0.5))):
+        offsets[vertex, axis] = offset
+    return cp.asarray(offsets)
```

Comment:
I wonder if `np.zeros()` is used here because calculating `offsets` in system memory and then converting them to GPU memory is more efficient than creating or manipulating a CuPy array.

Reply:
Yes, exactly. It will be orders of magnitude slower to launch kernels with only a few elements to process. It is best to just compute the offsets needed on the CPU and then transfer them.